docs/charms.reactive.decorators.rst

```diff
@@ -3,6 +3,38 @@ charms.reactive.decorators
 
 .. rubric:: Summary
 
+These decorators turn a regular function into a reactive handler so that it will
+be invoked if the conditions in the decorators are met. These decorators are
+also called the preconditions of the handlers. If a handler has multiple
+decorators, they will be ANDed together: the handler will only be invoked if
+all of the individual decorators match.
+
+Handlers are currently re-invoked for every hook execution, even if their
+predicate flags have not changed. However, this may well change in the future,
+and it is highly recommended that you not rely on this behavior. If you need to
+ensure that a particular handler runs again, you should set or remove one of its
+predicate flags.
+
+Regular handlers should not accept any arguments. When a handler needs to use a
+(relationship) :class:`~charms.reactive.altrelations.Endpoint`, it can access
+the endpoint object via the :data:`~charms.reactive.helpers.context` namespace.
+The only exceptions to this are enpoint handlers, handlers that are instance
+methods of :class:`~charms.reactive.altrelations.Endpoint`: they get the
+endpoint object as the `self` argment.
+
+If a handler needs to use an older interface using the legacy
+:class:`~charms.reactive.relations.RelationBase`, or if you're uncertain which
+is used by an interface layer, you can also obtain an instance using
```

Comment:
What do you mean with

>  if you're uncertain which is used by an interface layer

?

Reply:
Well, old-style relation instances aren't populated into the global context (because they depend on which specific flag you're handling), so you have to use `relation_factory` (or the terrible ambiguous handler args) to get access to them.  And whether an interface layer uses `EndpointInterface` or `RelationBase` is an implementation detail that the charm author shouldn't care about.  This will be problematic during the transition, so I'm not sure how to handle it gracefully.